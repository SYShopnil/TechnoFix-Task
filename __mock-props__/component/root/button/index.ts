import { BtnColorSchema, IRedirectButton } from '@src/types/root';

export const mock_data_redirectButton: IRedirectButton[] = [
  {
    btnText: 'Button Link',
    colorSchema: BtnColorSchema.SolidBgGrayTextViolet,
    isOpenNewTab: true,
    redirectLink: 'https://chat.openai.com/',
    dataTestId: 'Redirect Button One',
    isArrow: true,
  },
  {
    btnText: 'Back to Homepage Two',
    colorSchema: BtnColorSchema.SolidBgVioletTextWhite,
    isOpenNewTab: false,
    redirectLink: '',
    dataTestId: 'Redirect Button Two',
    isArrow: false,
  },
];
