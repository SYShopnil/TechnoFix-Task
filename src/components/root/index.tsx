export { IconStore } from './_icon';
export { Button } from './button';
export { DynamicCheckBox } from './dynamic-check-box';
