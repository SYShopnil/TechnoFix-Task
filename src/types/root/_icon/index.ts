export enum IconName {
  DownArrow = 'DownArrow',
  GraduationCap = 'GraduationCap',
  QuestionSign = 'QuestionSign',
}

export interface IIconStore {
  iconName: IconName;
  fill?: string;
}
