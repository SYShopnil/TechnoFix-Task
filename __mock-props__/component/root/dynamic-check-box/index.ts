import { IDynamicCheckBox } from '@src/types/root';

export const checkBoxItemsDataMock: {
  name: string;
  isCheck: boolean;
}[] = [
  {
    isCheck: true,
    name: 'title',
  },
  {
    isCheck: false,
    name: 'categories',
  },
  {
    isCheck: false,
    name: 'price',
  },
  {
    isCheck: true,
    name: 'date',
  },
  {
    isCheck: true,
    name: 'author',
  },
  {
    isCheck: true,
    name: 'status',
  },
  {
    isCheck: true,
    name: 'action',
  },
];
export const dynamicCheckBoxMock: IDynamicCheckBox = {
  checkBoxData: checkBoxItemsDataMock,
  checkBoxHandler: function (index) {
    console.log(index);
  },
  title: 'Add or removes columns',
  dataTestId: 'checkBoxGetter',
};
