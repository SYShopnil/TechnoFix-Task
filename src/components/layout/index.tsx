import { Button, IconStore } from '@src/components/root';
import { BtnColorSchema, IconEnums } from '@src/types/root';
import { ILandingPageLayout } from '@src/types/view';
import React from 'react';

export const Layout = ({ children, tableData }: ILandingPageLayout) => {
  return (
    <div className={`container`}>
      {/* header part */}
      <div
        className={`flex justify-between bg-[#FFFFFF] mt-5 p-4 shadow-md rounded-md`}
      >
        {/* title icon button side part wrapper */}
        <div className={`flex-[0_1_40%]`}>
          <div className={`flex justify-between`}>
            {/* icon and title part wrapper */}
            <div className={`flex justify-center items-center`}>
              <span className="flex-[0_1_10%] mr-8">
                <IconStore
                  iconName={IconEnums.GraduationCap}
                  fill={'#7A67ED'}
                />
              </span>
              <span
                className={`text-center flex-[0_1_90%] text-[1.2rem] font-semibold`}
              >
                All Courses
              </span>
            </div>
            <span className={`flex justify-center items-center`}>
              <Button.RedirectButton
                btnText="Add New Course"
                colorSchema={BtnColorSchema.SolidBgVioletTextWhite}
                isOpenNewTab={false}
                isArrow={false}
                redirectLink=""
              />
            </span>
          </div>
        </div>
        {/* fake wrapper */}
        <div className={`flex-[0_1_40%]`}></div>
        {/* question sign wrapper */}
        <div className={`flex-[0_1_20%] flex justify-center items-end`}>
          <div
            className={`bg-transparent border-[#7A67EE] border-2 border-solid  rounded-full p-2`}
          >
            <IconStore iconName={IconEnums.QuestionSign} fill={'#7A67ED'} />
          </div>
        </div>
      </div>
      <div>{children}</div>
    </div>
  );
};
