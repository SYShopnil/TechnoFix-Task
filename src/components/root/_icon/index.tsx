import { IIconStore, IconEnums } from '@src/types/root';
import React from 'react';

import { IoIosArrowDown } from 'react-icons/io';
import { RiGraduationCapFill } from 'react-icons/ri';
import { MdQuestionMark } from 'react-icons/md';

export const IconStore = ({ iconName, fill }: IIconStore) => {
  switch (iconName) {
    case IconEnums.DownArrow: {
      return <IoIosArrowDown />;
    }
    case IconEnums.GraduationCap: {
      return <RiGraduationCapFill fill={fill} size={55} />;
    }
    case IconEnums.QuestionSign: {
      return <MdQuestionMark fill={fill} size={45} />;
    }
    default: {
      return <IoIosArrowDown />;
    }
  }
};
