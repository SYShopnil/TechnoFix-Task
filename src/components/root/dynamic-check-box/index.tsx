import { IDynamicCheckBox } from '@src/types/root';
import React from 'react';

export const DynamicCheckBox = ({
  title,
  checkBoxData,
  checkBoxHandler,
  dataTestId,
}: IDynamicCheckBox) => {
  return (
    <div
      className={`bg-[#FFFFFF] shadow-primary p-4 space-y-4 `}
      data-testid={dataTestId || 'checkBoxPicker'}
    >
      <div>
        <h1 className={`font-semibold text-black`}>{title}</h1>
      </div>
      <div className={`space-y-8`}>
        {checkBoxData.map((item, ind) => {
          return (
            <div key={ind}>
              <label htmlFor={item.name}>
                <span className={`relative`}>
                  <input
                    type="checkbox"
                    onChange={() => checkBoxHandler(ind)}
                    checked={item.isCheck}
                    className={`appearance-none w-5 h-5 border-2 border-gray-300 rounded-md checked:bg-red-500 checked:border-red-600 checked:text-white checked:outline-none checked:relative checked:before:block checked:content-[''] checked:w-5 checked:h-5 checked:border-white checked:border-solid checked:mx-auto`}
                  />
                </span>
                <span
                  className={`ml-2 font-semibold capitalize text-[1.25rem]`}
                >
                  {item.name}
                </span>
              </label>
            </div>
          );
        })}
      </div>
    </div>
  );
};
