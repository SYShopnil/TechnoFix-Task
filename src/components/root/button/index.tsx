import {
  BtnColorSchema,
  IButton,
  INormalButton,
  IRedirectButton,
} from '@src/types/root';
import Link from 'next/link';
import React, { useState, useEffect } from 'react';
import { IconStore } from '../_icon';
import { IconName } from '@src/types/root/_icon';

//Common layout
const BaseButton = ({ btnText, colorSchema, isArrow }: IButton) => {
  const [fillValue, setFillValue] = useState('');
  const appliedBtnMainWrapperStyle = `flex justify-center items-center `;
  const appliedBtnInnerWrapperStyle = `text-base md:text-base_2  flex justify-center items-center px-[1rem] py-[0.5rem] rounded-[0.563rem]`;
  let appliedBtnBgColorStyle = '';

  const setRootFillValue = {
    initial: '',
    onHover: '',
  };
  switch (colorSchema) {
    case BtnColorSchema.SolidBgVioletTextWhite: {
      appliedBtnBgColorStyle = `bg-[#7A67EF] hover:bg-[transparent]  duration-[0.5s] text-[#FFFAFF] hover:text-[#7A67EF]  border-[0.063rem] border-primary_300 hover:border-[#7A67EF]`;
      setRootFillValue['onHover'] = '#727272';
      setRootFillValue['initial'] = '#FFFAFF';
      break;
    }
    case BtnColorSchema.SolidBgWhiteTextGreen: {
      appliedBtnBgColorStyle =
        'bg-[#EBF5EC] hover:bg-[#449657]  duration-[0.5s] text-[#449657] hover:text-[#EBF5EC]   border-[0.063rem] hover:border-[#EBF5EC] border-[#449657]';
      setRootFillValue['onHover'] = '#EBF5EC';
      setRootFillValue['initial'] = '#449657';
      break;
    }
    case BtnColorSchema.SolidBgGrayTextViolet: {
      appliedBtnBgColorStyle =
        'bg-[#EDEDED] hover:bg-[#7A67EE]  duration-[0.5s] text-[#7A67EE] hover:text-[#EDEDED]   border-[0.063rem] hover:border-[#EDEDED] border-[#7A67EE]';
      setRootFillValue['onHover'] = '#EDEDED';
      setRootFillValue['initial'] = '#7A67EE';
      break;
    }
    default: {
      appliedBtnBgColorStyle = `bg-[transparent] hover:bg-primary_500 duration-[0.5s] text-primary_500 hover:text-neutral_500 border-[0.063rem] border-primary_500`;
    }
  }

  // handle the first time arrow initialize
  useEffect(() => {
    setFillValue(setRootFillValue['initial']);
  }, []);

  return (
    // button main wrapper
    <span
      className={`${appliedBtnMainWrapperStyle} `}
      onMouseEnter={() => setFillValue(setRootFillValue.onHover)}
      onMouseLeave={() => setFillValue(setRootFillValue.initial)}
    >
      <span
        className={`${appliedBtnInnerWrapperStyle} w-[100%] h-[100%] ${appliedBtnBgColorStyle}   `}
      >
        <span>{btnText}</span>
        {/* btn text part wrapper */}
        {isArrow && (
          <span
            className={`ml-[1rem] h-[1.5rem] w-[1.5rem]  flex justify-center items-center`}
          >
            <IconStore iconName={IconName.DownArrow} fill={fillValue} />
          </span>
        )}
      </span>
    </span>
  );
};

//Normal Button
export const Button = (props: INormalButton) => {
  return (
    <span
      onClick={(
        e: React.MouseEvent<HTMLElement> | React.FormEvent<HTMLFormElement>
      ) => props.clickHandler?.(e)}
      data-testid={props.dataTestId}
    >
      <BaseButton {...props} />
    </span>
  );
};

//Redirect Button
Button.RedirectButton = (props: IRedirectButton) => {
  return (
    <Link href={props.redirectLink} passHref>
      <a
        target={props.isOpenNewTab ? '_blank' : '_self'}
        data-testid={props.dataTestId}
      >
        <BaseButton {...props} />
      </a>
    </Link>
  );
};
