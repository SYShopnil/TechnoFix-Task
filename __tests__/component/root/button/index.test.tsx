/**
 *  DONE
 *
 *  Check render successfully of not by taking a snapshot
 *  Check the redirect function work perfectly or not
 *  Check Open new tab function work perfectly or not
 *
 */

import { mock_data_redirectButton } from '@root/__mock-props__/component/root/button';
import { Button } from '@src/components/root';
import { render } from '@testing-library/react';

let buttonCounter = 0;

// Mock the 'next/link' module
jest.mock('next/link', () => {
  // Require the actual implementation to mock the 'Link' component correctly
  const nextLink = jest.requireActual('next/link');

  return {
    __esModule: true,
    default: ({ children }: { children: React.ReactNode }) => children,
    // Exporting the original components from 'next/link' if needed
    ...nextLink,
  };
});

describe.each(mock_data_redirectButton)(
  `When Button ${++buttonCounter} component has rendered`,
  (buttons) => {
    it(`Expect it has been rendered successfully`, () => {
      const { getByTestId } = render(<Button.RedirectButton {...buttons} />);
      const buttonElement = getByTestId(buttons.dataTestId || '');
      expect(buttonElement).toBeInTheDocument();
      expect(buttonElement).toMatchSnapshot();
    });

    it(`Expect redirect function worked perfectly`, () => {
      const { getByTestId } = render(<Button.RedirectButton {...buttons} />);
      const buttonElement = getByTestId(buttons.dataTestId || '');

      if (buttons.redirectLink) {
        expect(buttonElement).toHaveAttribute('href', buttons.redirectLink);
      } else {
        expect(buttonElement).toHaveAttribute('href', '/');
      }
    });

    it(`Expect open new tab function worked perfectly`, () => {
      const { getByTestId } = render(<Button.RedirectButton {...buttons} />);
      const buttonElement = getByTestId(buttons.dataTestId || '');

      if (buttons.isOpenNewTab) {
        expect(buttonElement).toHaveAttribute('target', '_blank');
      } else {
        expect(buttonElement).toHaveAttribute('target', '_self');
      }
    });
  }
);
