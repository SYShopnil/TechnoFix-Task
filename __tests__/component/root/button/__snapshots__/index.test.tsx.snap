// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When Button 1 component has rendered Expect it has been rendered successfully 1`] = `
<a
  data-testid="Redirect Button One"
  href="https://chat.openai.com/"
  target="_blank"
>
  <span
    class="flex justify-center items-center  "
  >
    <span
      class="text-base_2  flex justify-center items-center px-[1.5rem] py-[1rem] rounded-[0.563rem] w-[100%] h-[100%] bg-[#EDEDED] hover:bg-[#7A67EE]  duration-[0.5s] text-[#7A67EE] hover:text-[#EDEDED]   border-[0.063rem] hover:border-[#EDEDED] border-[#7A67EE]  "
    >
      <span>
        Button Link
      </span>
      <span
        class="ml-[1rem] h-[1.5rem] w-[1.5rem]  flex justify-center items-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
          />
        </svg>
      </span>
    </span>
  </span>
</a>
`;

exports[`When Button 1 component has rendered Expect it has been rendered successfully 2`] = `
<a
  data-testid="Redirect Button Two"
  href="/"
  target="_self"
>
  <span
    class="flex justify-center items-center  "
  >
    <span
      class="text-base_2  flex justify-center items-center px-[1.5rem] py-[1rem] rounded-[0.563rem] w-[100%] h-[100%] bg-[#7A67EF] hover:bg-[transparent]  duration-[0.5s] text-[#FFFAFF] hover:text-[#7A67EF]  border-[0.063rem] border-primary_300 hover:border-[#7A67EF]  "
    >
      <span>
        Back to Homepage Two
      </span>
    </span>
  </span>
</a>
`;
